spark-cassandra-connector-embedded/src/main/scala/com/datastax/spark/connector/embedded/EmbeddedCassandra.scala

```diff
@@ -43,60 +46,81 @@ trait EmbeddedCassandra {
 }
 
 object UserDefinedProperty {
-  abstract class NodeProperty(val propertyName: String)
-  type PropertyType = NodeProperty {val propertyName: String}
-  case object HostProperty extends NodeProperty("IT_TEST_CASSANDRA_HOSTS")
-  case object NativePortProperty extends NodeProperty("IT_TEST_CASSANDRA_NATIVE_PORTS")
-  case object RpcPortProperty extends NodeProperty("IT_TEST_CASSANDRA_RPC_PORTS")
-
-  val hosts = getNodePropertySeq(HostProperty).asInstanceOf[IndexedSeq[InetAddress]]
-  val nativePorts = getNodePropertySeq(NativePortProperty).asInstanceOf[IndexedSeq[Int]]
-  val rpcPorts = getNodePropertySeq(RpcPortProperty).asInstanceOf[IndexedSeq[Int]]
-
-  private def getNodePropertySeq(nodeProperty: NodeProperty): IndexedSeq[Any] = {
-    nodeProperty match {
-      case NativePortProperty => getValueSeq(NativePortProperty.propertyName).map(e => e.toInt)
-      case RpcPortProperty => getValueSeq(RpcPortProperty.propertyName).map(e => e.toInt)
-      case HostProperty => getValueSeq(HostProperty.propertyName).map(e => InetAddress.getByName(e))
-      case _ => throw new RuntimeException("Wrong node input property")
-    }
+
+  trait TypedProperty {
+    type ValueType
+    def convertValueFromString(str: String): ValueType
+    def checkValueType(obj: Any): ValueType
+  }
+
+  trait IntProperty extends TypedProperty {
+    type ValueType = Int
+    def convertValueFromString(str: String) = str.toInt
+    def checkValueType(obj: Any) =
+      obj match {
+        case x: Int => x
+        case _ => throw new ClassCastException (s"Expected Int, got ${obj.getClass.getName}")
+      }
+  }
+
+  trait InetAddressProperty extends TypedProperty {
+    type ValueType = InetAddress
+    def convertValueFromString(str: String) = InetAddress.getByName(str)
+    def checkValueType(obj: Any) =
+      obj match {
+        case x: InetAddress => x
+        case _ => throw new ClassCastException (s"Expected InetAddress, got ${obj.getClass.getName}")
```

Comment:
I believe I was thinking of an apply for whatever the implementer of InetAddressProperty would be, because I saw something invoking 'checkValueType'. But now that I think of it, 'checkValueType' is more an 'unapply' behavior.